⚡ feat: bump minor version to add support for APS environment variables

MadhukarMoogala opened the pull request:
⚡ **feat**: bump minor version to add support for APS environment variables

- Introduced `AddAPSAlternativeEnvironmentVariables` extension method to add APS environment variables to the configuration builder.
- Created `APSAlternativeConfigurationSource` class to represent the configuration source for APS.
- Implemented `APSAlternativeConfigurationProvider` class to load APS configuration from environment variables.

Add unit tests for APS alternative environment variables

- Added `TestValuesFromAPSEnvironment` to verify loading of APS environment variables.
- Ensured tests cover various configuration sources including environment variables, JSON, and legacy environment variables.

zhuliice commented on src/Autodesk.Forge.Core/Autodesk.Forge.Core.csproj:
```diff
@@ -3,6 +3,7 @@
   <PropertyGroup>
     <Description>Shared code for APS client sdks</Description>
     <GenerateDocumentationFile>true</GenerateDocumentationFile>
+    <Version>4.1.1</Version>
```
the version should be updated at: https://github.com/Autodesk-Forge/forge-api-dotnet-core/blob/b5b4065b6b7243106f326b47f0d8251efbd886b4/Directory.Build.props#L3

MadhukarMoogala replied:
Thanks, I will add.

zhuliice commented on CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+### 4.1.1
+
+- Add support for APS alternative environment variables `APS_CLIENT_ID` and `APS_CLIENT_SECRET`
+
```
could you also state that `FORGE_CLIENT_ID` and `FORGE_CLIENT_SECRET` environment variables are not encouraged to use anymore

zhuliice commented on src/Autodesk.Forge.Core/LegacySampleConfigurationProvider.cs:
```diff
@@ -74,5 +89,49 @@ public override void Load()
+            var secret = Environment.GetEnvironmentVariable("APS_CLIENT_SECRET");
+            if (!string.IsNullOrEmpty(secret))
+            {
+                this.Data.Add("Forge:ClientSecret", secret);
```
I think we should cover the code here too:

[forge-api-dotnet-core/src/Autodesk.Forge.Core/ServiceCollectionExtensions.cs](https://github.com/Autodesk-Forge/forge-api-dotnet-core/blob/b5b4065b6b7243106f326b47f0d8251efbd886b4/src/Autodesk.Forge.Core/ServiceCollectionExtensions.cs#L39)

Line 39 in [b5b4065](https://github.com/Autodesk-Forge/forge-api-dotnet-core/commit/b5b4065b6b7243106f326b47f0d8251efbd886b4)

 services.Configure<ForgeConfiguration>(configuration.GetSection("Forge")); 
to allow section named as APS

so `this.Data.Add("Forge:ClientId", id);` should be `this.Data.Add("APS:ClientId", id);`

otherwise it is confusing to me that we still HAVE TO define `Forge` in appsettings.json
```
"Forge": {
        "ClientId": "<default clientId>",
        "ClientSecret": "<default clientSecret>",
```

MadhukarMoogala replied:
Thanks, I will add that.

I'm thinking like this to add another configure without removing Forge.
Should we remove Forge section entirely, that may be a breaking change update.

```
            services.Configure<ForgeConfiguration>(configuration.GetSection("Forge"));
            services.Configure<ForgeConfiguration>(configuration.GetSection("APS"));
```

szilvaa-adsk replied:
Yep. Let's not break existing clients quite yet. Give them few years to adapt.

zhuliice commented on tests/Autodesk.Forge.Core.Test/TestForgeConfiguration.cs:
```diff
@@ -51,6 +51,23 @@ public void TestValuesFromLegacyEnvironment()
+            Environment.SetEnvironmentVariable("APS_CLIENT_ID", "bla");
+            Environment.SetEnvironmentVariable("APS_CLIENT_SECRET", "blabla");
+            var configuration = new ConfigurationBuilder()
+                .AddAPSAlternativeEnvironmentVariables()
```
what if I call `AddForgeAlternativeEnvironmentVariables` too?

szilvaa-adsk replied:
We should mark the old stuff with `[Obsolete]` so you get a warning from the compiler.

MadhukarMoogala replied:
If you call that it works too, but I will mark it as [Obsolete].

Thanks.

zhuliice commented on Directory.Build.props:
```diff
@@ -1,6 +1,6 @@
 <Project>
  <PropertyGroup>
-   <Version>4.0.1</Version>
+   <Version>4.1.1</Version>
```
`4.1.0`?

MadhukarMoogala replied:
ah! got.  thanks.

MadhukarMoogala commented:
### Summary
This PR introduces APS configuration handling with new unit tests, alternative APS environment variable support.

### Changes
- **Unit Tests for APS Configuration:**
  - `TestAPSConfigFromEnvironmentVariables_DoubleUnderscoreFormat`: Tests loading of APS environment variables in double underscore format.
  - `TestAPSConfigFromEnvironmentVariables_UnderscoreFormat`: Tests loading of APS environment variables in underscore format.
  - `TestAPSConfigFromJson`: Tests loading APS configuration from JSON.
  - `TestAPSConfigFromJsonWithAgents`: Tests loading APS configuration with additional agents from JSON.
  - `TestAPSUserAgent`: Verifies APS configuration for user agent "user1" with authentication and request handling.

- **APS Environment Variable Support:**
  - Added `AddAPSAlternativeEnvironmentVariables` to support APS-specific environment variables.
  - Implemented `APSAlternativeConfigurationSource` and `APSAlternativeConfigurationProvider` to load APS settings from environment variables.
  - Marked `AddForgeAlternativeEnvironmentVariables` as obsolete for transition to APS naming.

- **Service Collection Enhancements:**
  - Updated `AddForgeService` to support APS configurations in both "Forge" and "APS" sections, ensuring backward compatibility.
 

This PR ensures smooth migration from Forge to APS naming conventions.

zhuliice commented on src/Autodesk.Forge.Core/Autodesk.Forge.Core.csproj:
```diff
@@ -3,6 +3,7 @@
   <PropertyGroup>
     <Description>Shared code for APS client sdks</Description>
     <GenerateDocumentationFile>true</GenerateDocumentationFile>
+    <Version>4.1.1</Version>
```
can you remove it?

zhuliice commented on src/Autodesk.Forge.Core/ServiceCollectionExtensions.cs:
```diff
@@ -37,6 +37,7 @@ public static IHttpClientBuilder AddForgeService(this IServiceCollection service
         {
             services.AddOptions();
             services.Configure<ForgeConfiguration>(configuration.GetSection("Forge"));
```
@szilvaa is there some way to get a warning on this too? e.g. if the user still defines `Forge` in appsettings.json

szilvaa-adsk replied:
We would need to add our own `.targets` file to the nuget package that would check the user appsettings.json to see if they still use this old stuff. So, it is possible, but it isn't trivial.

zhuliice reviewed (approved):
thanks for the work!
